File: app/src/main/java/sk/styk/martin/apkanalyzer/business/task/LocalStatisticsLoader.java
package sk.styk.martin.apkanalyzer.business.task;

import android.content.Context;

import java.lang.ref.WeakReference;
import java.util.List;

import sk.styk.martin.apkanalyzer.business.service.AppBasicDataService;
import sk.styk.martin.apkanalyzer.business.service.LocalApplicationStatisticDataService;
import sk.styk.martin.apkanalyzer.model.statistics.LocalStatisticsData;
import sk.styk.martin.apkanalyzer.model.statistics.LocalStatisticsDataBuilder;

public class LocalStatisticsLoader extends ApkAnalyzerAbstractAsyncLoader<LocalStatisticsData> {
    public static final int ID = 3;

    private WeakReference<ProgressCallback> callbackReference;
    private AppBasicDataService appBasicDataService;
    private LocalApplicationStatisticDataService localAppDataService;

    public interface ProgressCallback {
        void onProgressChanged(int currentProgress, int maxProgress);
    }

    public LocalStatisticsLoader(Context context, ProgressCallback callback) {
        super(context);
        callbackReference = new WeakReference<>(callback);
        appBasicDataService = new AppBasicDataService(context.getPackageManager());
        localAppDataService = new LocalApplicationStatisticDataService(context.getPackageManager());
    }

    public void setCallbackReference(ProgressCallback progressCallback){
        callbackReference = new WeakReference<>(progressCallback);
    }

    @Override
    public LocalStatisticsData loadInBackground() {
        List<String> packageNames = appBasicDataService.getAllPackageNames();
        LocalStatisticsDataBuilder builder = new LocalStatisticsDataBuilder(packageNames.size());

        for (int i = 0; i < packageNames.size(); i++) {
            builder.add(localAppDataService.get(packageNames.get(i)));
            if (callbackReference.get() != null) {
                callbackReference.get().onProgressChanged(i, packageNames.size());
            }
        }

        return builder.build();
    }

}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/task/LocalStatisticsLoader.kt
package sk.styk.martin.apkanalyzer.business.task

import android.content.Context
import sk.styk.martin.apkanalyzer.business.service.launcher.AppBasicDataService
import sk.styk.martin.apkanalyzer.business.service.launcher.LocalApplicationStatisticDataService
import sk.styk.martin.apkanalyzer.model.statistics.LocalStatisticsData
import sk.styk.martin.apkanalyzer.model.statistics.LocalStatisticsDataBuilder
import java.lang.ref.WeakReference

/**
 * @author Martin Styk
 * @version 15.09.2018
 */
class LocalStatisticsLoader(context: Context, callback: ProgressCallback) : ApkAnalyzerAbstractAsyncLoader<LocalStatisticsData>(context) {

    private var callbackReference = WeakReference(callback)
    private val appBasicDataService = AppBasicDataService(context.packageManager)
    private val localAppDataService = LocalApplicationStatisticDataService(context.packageManager)

    interface ProgressCallback {
        fun onProgressChanged(currentProgress: Int, maxProgress: Int)
    }

    fun setCallbackReference(progressCallback: ProgressCallback) {
        callbackReference = WeakReference(progressCallback)
    }

    override fun loadInBackground(): LocalStatisticsData {
        val packageNames = appBasicDataService.getAllPackageNames()
        val builder = LocalStatisticsDataBuilder(packageNames.size)

        for (i in packageNames.indices) {
            builder.add(localAppDataService.get(packageNames[i]))

            callbackReference.get()?.onProgressChanged(i, packageNames.size)
        }

        return builder.build()
    }

    companion object {
        const val ID = 3
    }

}

